src/tests/scroll_factor.rs

```diff
@@ -0,0 +1,315 @@
+use niri_config::{FloatOrInt, Mouse, Touchpad};
+
+// Reproduce the macro from input/mod.rs for testing
+macro_rules! extract_scroll_factors {
+    ($cfg:expr) => {{
+        let h_factor = $cfg
+            .scroll_factor_horizontal
+            .map(|x| x.0)
+            .or_else(|| $cfg.scroll_factor.map(|x| x.0))
+            .unwrap_or(1.);
+        let v_factor = $cfg
+            .scroll_factor_vertical
+            .map(|y| y.0)
+            .or_else(|| $cfg.scroll_factor.map(|y| y.0))
+            .unwrap_or(1.);
+        (h_factor, v_factor)
+    }};
+}
+
+fn create_mouse_config(
+    scroll_factor: Option<f64>,
+    scroll_factor_horizontal: Option<f64>,
+    scroll_factor_vertical: Option<f64>,
+) -> Mouse {
+    Mouse {
+        off: false,
+        natural_scroll: false,
+        accel_speed: Default::default(),
+        accel_profile: None,
+        scroll_method: None,
+        scroll_button: None,
+        scroll_button_lock: false,
+        left_handed: false,
+        middle_emulation: false,
+        scroll_factor: scroll_factor.map(FloatOrInt),
+        scroll_factor_horizontal: scroll_factor_horizontal.map(FloatOrInt),
+        scroll_factor_vertical: scroll_factor_vertical.map(FloatOrInt),
+    }
+}
+
+fn create_touchpad_config(
+    scroll_factor: Option<f64>,
+    scroll_factor_horizontal: Option<f64>,
+    scroll_factor_vertical: Option<f64>,
+) -> Touchpad {
+    Touchpad {
+        off: false,
+        tap: false,
+        dwt: false,
+        dwtp: false,
+        drag: None,
+        drag_lock: false,
+        natural_scroll: false,
+        click_method: None,
+        accel_speed: Default::default(),
+        accel_profile: None,
+        scroll_method: None,
+        scroll_button: None,
+        scroll_button_lock: false,
+        tap_button_map: None,
+        left_handed: false,
+        disabled_on_external_mouse: false,
+        middle_emulation: false,
+        scroll_factor: scroll_factor.map(FloatOrInt),
+        scroll_factor_horizontal: scroll_factor_horizontal.map(FloatOrInt),
+        scroll_factor_vertical: scroll_factor_vertical.map(FloatOrInt),
+    }
+}
+
+#[test]
```

Comment:
I very much appreciate the testing, but I'm not sure this fairly simple logic needs such an array of tests?

Reply:
No problem; I've simplified the battery of tests to just the most important ones. :smile: